libsolidity/codegen/ir/IRGeneratorForStatements.cpp

```diff
@@ -1063,6 +1055,77 @@ void IRGeneratorForStatements::endVisit(FunctionCall const& _functionCall)
 
 		break;
 	}
+	case FunctionType::Kind::ECRecover:
+	case FunctionType::Kind::RIPEMD160:
+	case FunctionType::Kind::SHA256:
+	{
+		solAssert(!_functionCall.annotation().tryCall, "");
+		solAssert(!functionType->valueSet(), "");
+		solAssert(!functionType->gasSet(), "");
+		solAssert(!functionType->bound(), "");
+
+		static map<FunctionType::Kind, std::tuple<u160, size_t>> precompiles = {
+			{FunctionType::Kind::ECRecover, std::make_tuple(1, 0)},
+			{FunctionType::Kind::SHA256, std::make_tuple(2, 0)},
+			{FunctionType::Kind::RIPEMD160, std::make_tuple(3, 12)},
+		};
+		auto [ address, offset ] = precompiles[functionType->kind()];
+		TypePointers argumentTypes;
+		vector<string> argumentStrings;
+		for (auto const& arg: arguments)
+			argumentTypes.emplace_back(&type(*arg));
+		for (auto const& arg: arguments | boost::adaptors::reversed)
+			if (IRVariable(*arg).type().sizeOnStack() > 0)
+				argumentStrings.emplace_back(IRVariable(*arg).commaSeparatedList());
+		string argumentString = argumentStrings.empty() ? ""s : (", " + joinHumanReadable(argumentStrings));
+		ABIFunctions abi(m_context.evmVersion(), m_context.revertStrings(), m_context.functionCollector());
+		Whiskers templ(R"(
+			let <pos> := <freeMemory>
+			let <end> := <encodeArgs>(<pos> <argumentString>)
+
+			<?isECRecover>
+				mstore(0, 0)
+			</isECRecover>
+
+			let <success> := <call>(<gas>, <address> <?isCall>, 0</isCall>, <pos>, sub(<end>, <pos>), 0, 32)
+
+			<?isECRecover>
+			<!isECRecover>
+				if iszero(<success>) { <forwardingRevert>() }
+			</isECRecover>
+
+			let <retVars> := <shl>(mload(0))
+		)");
+		templ("call", m_context.evmVersion().hasStaticCall() ? "staticcall" : "call");
+		templ("isCall", !m_context.evmVersion().hasStaticCall());
+		templ("shl", m_utils.shiftLeftFunction(offset * 8));
+		templ("freeMemory", freeMemory());
+		templ("pos", m_context.newYulVariable());
+		templ("end", m_context.newYulVariable());
+		templ("isECRecover", FunctionType::Kind::ECRecover == functionType->kind());
+		if (FunctionType::Kind::ECRecover == functionType->kind())
+			templ("encodeArgs", abi.tupleEncoder(argumentTypes, parameterTypes));
+		else
+			templ("encodeArgs", abi.tupleEncoderPacked(argumentTypes, parameterTypes));
+		templ("argumentString", argumentString);
+		templ("address", toString(address));
+		templ("success", m_context.newYulVariable());
+		templ("retVars", IRVariable(_functionCall).commaSeparatedList());
+		templ("forwardingRevert", m_utils.forwardingRevertFunction());
+		if (m_context.evmVersion().canOverchargeGasForCall())
+			// Send all gas (requires tangerine whistle EVM)
+			templ("gas", "gas()");
+		else
+		{
+			u256 gasNeededByCaller = evmasm::GasCosts::callGas(m_context.evmVersion()) + 3 +
+										evmasm::GasCosts::callNewAccountGas;
```

Comment:
I just use `callNewAccountGas` here, but I guess it's somehow misleading.

Reply:
What is the `+10` ?

Reply:
I had the same question. @chriseth told me `+10` is because of `gas` and `sub`. 

I was now looking into the gas prices. The pricing tier for `gas` seem to be _base_, where `sub` seem to be _verylow_. That makes `gas` to be `2` and `sub` to be `3`. I'm not sure now where the `+10` is exactly coming from.

In some experiments I saw that `+3` would also do the job here. But the old code-generator used `+10` so I thought that its better to not change this.

Reply:
Right, this is for the ancient EVM case.

Reply:
`PUSH <gasNeededByCaller> GAS SUB` is the exact assembly generated. Plus I guess some overhead to leave space justifies the 10.

The old code also said "need to fine-tune":
```
                // send all gas except the amount needed to execute "SUB" and "CALL"
                // @todo this retains too much gas for now, needs to be fine-tuned.
                u256 gasNeededByCaller = evmasm::GasCosts::callGas(m_context.evmVersion()) + 10;
                if (_functionType.valueSet())
                        gasNeededByCaller += evmasm::GasCosts::callValueTransferGas;
                if (!existenceChecked)
                        gasNeededByCaller += evmasm::GasCosts::callNewAccountGas; // we never know
                m_context << gasNeededByCaller << Instruction::GAS << Instruction::SUB;
```

I would say leave a comment "@todo The value 10 is not exact and this could be fine-tuned, but this has worked for years in the old code generator."